bionic/core/task_execution.py

```diff
@@ -313,16 +278,21 @@ class EntryRequirement:
     """
     Represents a requirement from one entry to another.
 
-    A requirement indicates that one entry (`src_entry`) can't make any further progress
-    until another entry (`dst_entry`) has reached a certain level of progress (`level`).
+    A requirement indicates that we need a particular entry (``dst_entry``) to reach a
+    certain level of progress (``level``). This could be because it's blocking some
+    other entry (``src_entry``), or for some external or a priori reason.
 
-    Note: `src_entry` can also be `None`, indicating some sort of external or a priori
-    requirement. On the other hand, `dst_entry` cannot be `None`.
+    An entry can be marked ``is_transient``, meaning it can safely be discarded once
+    ``dst_entry`` has reached the target ``level``. Otherwise, it also be discarded if
+    its ``src_entry`` has reached the (final) ``CACHED`` level. Once a requirement is
+    discarded, it's permissible for ``dst_entry`` to revert to a previous level. If a
+    requirement is not transient and has no source, it is assumed to be permanent.
     """
 
-    src_entry = attr.ib()
-    dst_entry = attr.ib()
-    level = attr.ib()
+    dst_entry: TaskRunnerEntry = attr.ib()
+    level: "EntryLevel" = attr.ib()
+    src_entry: Optional[TaskRunnerEntry] = attr.ib()
+    is_transient: bool = attr.ib()
```

Comment:
This might be only me, but I find `transient` hard to understand since it isn't clear when it disappears. It seems like you are using `transient` to tell if only `dst` cares about the requirement. When it is not transient, it means that either `src` cares about it, or the `dst` was requested (so deriver cares about it). How do you feel about encoding these two properties separately instead of combining them in one property (which is `is_transient`)?

Reply:
As we discussed in a hangout, I'll try replacing this with some kind of `valid_until` enum.

Reply:
I replaced `is_transient` with an `expiration` enum field.

Reply:
Thanks! I find the code a lot easier to read now.